graphbolt/src/fused_csc_sampling_graph.cc

```diff
@@ -335,13 +335,12 @@ FusedCSCSamplingGraph::SampleNeighborsImpl(
         auto num_picked_neighbors_data_ptr =
             num_picked_neighbors_per_node.data_ptr<scalar_t>();
         num_picked_neighbors_data_ptr[0] = 0;
-        const auto nodes_data_ptr = nodes.data_ptr<int64_t>();
 
         // Step 1. Calculate pick number of each node.
         torch::parallel_for(
             0, num_nodes, grain_size, [&](int64_t begin, int64_t end) {
               for (int64_t i = begin; i < end; ++i) {
-                const auto nid = nodes_data_ptr[i];
+                const auto nid = nodes[i].item<int64_t>();
```

Comment:
I think this might have a negative effect on runtime if it uses dynamic polymorphism, I would have preferred casting nodes to the same dtype as indices before calling this function. Or utilizing AT_DISPATCH_INTEGRAL_TYPES for nodes as well.

Reply:
I though about it but I don't think we really need to do like this. And I think it's `static_cast<>` behind `item<>`. No overhead in runtime. We do need to make sure the underlying data type could be safely casted to the `int64_t`, but such assumption is also required even using `scalar_t`. What's more, if the underlying dtype is float, `const auto nid = scalar_t` returns float while `item<>` casts it to `int64_t`.

Reply:
The problem is that with the `.item` member, the code needs to look at the underlying tensor type information each time it is called (inside of the for loop), go into the corresponding branch in the code, and then return the value. (This is performed for every access.)

Reply:
This is basically a (minor, should be benchmarked) performance concern, the code is correct of course.

Reply:
Thanks for your suggestion.

Reply:
I haven't checked the implementation of `item<>`. But I guess `nodes[i].item<int64_t>()` is like
```
static_cast<int64_t>(nodes.data_ptr<scalar_t>()[i])
```
Is this workable?

Reply:
Sorry, it is my mistake. I was slightly mistaken about the above description. Let me correct it below, the performance issue is basically due to the `nodes[i]` access:

https://github.com/pytorch/pytorch/blob/d70857bd9e425ffc5b1a32639fb6651462eeef97/aten/src/ATen/templates/TensorBody.h#L321 has the definition of the tensor::operator[], which is being used in `nodes[i].item<int64_t>();`. It calls the tensor::select function. Which is defined here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorShape.cpp#L1816.

I am writing it for referencing and self-correcting.

Reply:
The array access needs to happen after C++ knows the type information, which means we need to access it via a pointer whose type C++ already knows, which means we have to unfortunately use a dispatch for maximum performance especially because it is performed in a for loop.

Reply:
I hope I made my point clear, sorry for the confusing explanation, I am not the best communicator.

Reply:
I see. So point is we'd better always prefer `nodes.data_ptr<sclar_t>()[i]` than `nodes[i]` for indexing on tensor as the latter one requires to call `select()` each time while the previous one does not as we have already accessing on the underlying data pointer directly. Right?

Reply:
Exactly, and `nodes.data_ptr<scalar_t>()` unfortunately is inconvenient and requires a dispatch to get `scalar_t`. But this way, we pay the dispatch cost outside any loops.